### PR TITLE
added transactions hashes and execution order in block structure

### PR DESCRIPTION
**Pull request by miiu96**
- Added transactions execution order in the `Block` structure.
- Added transactions hashes hex-encoded in the `MiniblockDetails` structure from the `Block` strucutre

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,7 +3,7 @@ module github.com/ElrondNetwork/elastic-indexer-go
 go 1.17
 
 require (
-	github.com/ElrondNetwork/elrond-go-core v1.1.26
+	github.com/ElrondNetwork/elrond-go-core v1.1.27-0.20221215101151-0e69a10c3cb5
```
Comment (iulianpascalau) [inline go.mod:6]: proper release

File: process/elasticproc/block/blockProcessor.go [hunk 2]
```diff
@@ -288,3 +327,26 @@ func createShardIdentifier(shardID uint32) uint32 {
 func (bp *blockProcessor) ComputeHeaderHash(header coreData.HeaderHandler) ([]byte, error) {
 	return core.CalculateHash(bp.marshalizer, bp.hasher, header)
 }
+
+func getTxsMap(mbType nodeBlock.Type, pool *outport.Pool) map[string]coreData.TransactionHandlerWithGasUsedAndFee {
+	switch mbType {
```
Comment (iulianpascalau) [inline process/elasticproc/block/blockProcessor.go:332]: will definitely need to refactor these maps so we can get rid of this type of switch. 
In another PR

File: process/elasticproc/block/blockProcessor.go [hunk 3]
```diff
@@ -227,17 +230,53 @@ func (bp *blockProcessor) getEncodedMBSHashes(body *block.Body) []string {
...
+	indexOfFirstProcessed, indexOfLastProcessed := mbHeader.GetIndexOfFirstTxProcessed(), mbHeader.GetIndexOfLastTxProcessed()
+	for idx, txHash := range txsHashes {
+		tx, found := txsMap[string(txHash)]
+		if !found {
+			log.Warn("blockProcessor.extractExecutionOrderIndicesFromPool cannot find tx in pool", "txHash", hex.EncodeToString(txHash))
+			continue
+		}
+
+		isExecutedInCurrentBlock := int32(idx) >= indexOfFirstProcessed && int32(idx) <= indexOfLastProcessed
+		if !isExecutedInCurrentBlock {
```
Comment (AdoAdoAdo) [inline process/elasticproc/block/blockProcessor.go:269]: can this be moved on line 262 ?
it seems you don't need the tx for the not executed txs in a partially executed mb, so you shouldn't return a warning.

File: process/elasticproc/block/blockProcessor_test.go [hunk 4]
```diff
@@ -274,3 +288,132 @@ func TestBlockProcessor_PrepareBlockForDBEpochStartMeta(t *testing.T) {
...
+			}},
+		Scrs: map[string]coreData.TransactionHandlerWithGasUsedAndFee{
+			scrHash: &outport.TransactionHandlerWithGasAndFee{
+				ExecutionOrder: 0,
+			},
+		},
+	})
+	require.Nil(t, err)
+
+	require.Equal(t, &data.Block{
```
Comment (AdoAdoAdo) [inline process/elasticproc/block/blockProcessor_test.go:371]: it is hard to follow that the expected value is correct, maybe think of a way to refactor in a future PR (e.g create a function to return expected block based on an input)

File: process/elasticproc/block/blockProcessor_test.go [hunk 5]
```diff
@@ -274,3 +288,132 @@ func TestBlockProcessor_PrepareBlockForDBEpochStartMeta(t *testing.T) {
...
+	bp, _ := NewBlockProcessor(&mock.HasherMock{}, &mock.MarshalizerMock{})
+
+	mbhr := &dataBlock.MiniBlockHeaderReserved{
+		IndexOfFirstTxProcessed: 0,
+		IndexOfLastTxProcessed:  0,
+	}
+	mbhrBytes, _ := gogoMarshaller.Marshal(mbhr)
+
+	txHash, notExecutedTxHash, invalidTxHash, rewardsTxHash, scrHash := "tx", "notExecuted", "invalid", "reward", "scr"
+	dbBlock, err := bp.PrepareBlockForDB([]byte("hash"), &dataBlock.Header{
```
Comment (AdoAdoAdo) [inline process/elasticproc/block/blockProcessor_test.go:305]: this is a bit heavy, maybe you can think of a way to refactor in a separate PR.

File: process/elasticproc/block/blockProcessor.go [hunk 6]
```diff
@@ -227,17 +230,53 @@ func (bp *blockProcessor) getEncodedMBSHashes(body *block.Body) []string {
...
 }
 
+func extractExecutionOrderIndicesFromPool(mbHeader coreData.MiniBlockHeaderHandler, txsHashes [][]byte, pool *outport.Pool) []int {
+	txsMap := getTxsMap(nodeBlock.Type(mbHeader.GetTypeInt32()), pool)
+	executionOrderTxsIndices := make([]int, len(txsHashes))
+	indexOfFirstProcessed, indexOfLastProcessed := mbHeader.GetIndexOfFirstTxProcessed(), mbHeader.GetIndexOfLastTxProcessed()
+	for idx, txHash := range txsHashes {
+		tx, found := txsMap[string(txHash)]
+		if !found {
+			log.Warn("blockProcessor.extractExecutionOrderIndicesFromPool cannot find tx in pool", "txHash", hex.EncodeToString(txHash))
```
Comment (AdoAdoAdo) [inline process/elasticproc/block/blockProcessor.go:264]: for the not found Tx it will appear as the order is 0, can we mark it as not found?
e.g create a const txNotFound = -2 ?

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (miiu96) [reply go.mod:6]: will do

File: process/elasticproc/block/blockProcessor.go (hunk 2 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor.go:332]: In the next refactor it will be fixed.

File: process/elasticproc/block/blockProcessor.go (hunk 3 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor.go:269]: moved to line 262. and removed the `warn` log

File: process/elasticproc/block/blockProcessor.go (hunk 6 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor.go:264]: added

File: process/elasticproc/block/blockProcessor_test.go (hunk 4 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor_test.go:371]: I will add a task for this refactor of the tests.

File: process/elasticproc/block/blockProcessor_test.go (hunk 5 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor_test.go:305]: I will add a task for this refactor of the tests.

File: process/elasticproc/block/blockProcessor.go [hunk 7]
```diff
@@ -227,17 +230,53 @@ func (bp *blockProcessor) getEncodedMBSHashes(body *block.Body) []string {
...
+			TxsHashes:                hexEncodeSlice(txsHashes),
+			ExecutionOrderTxsIndices: extractExecutionOrderIndicesFromPool(mbHeader, txsHashes, pool),
 		})
 	}
 }
 
+func extractExecutionOrderIndicesFromPool(mbHeader coreData.MiniBlockHeaderHandler, txsHashes [][]byte, pool *outport.Pool) []int {
+	txsMap := getTxsMap(nodeBlock.Type(mbHeader.GetTypeInt32()), pool)
+	executionOrderTxsIndices := make([]int, len(txsHashes))
+	indexOfFirstProcessed, indexOfLastProcessed := mbHeader.GetIndexOfFirstTxProcessed(), mbHeader.GetIndexOfLastTxProcessed()
```
Comment (SebastianMarian) [inline process/elasticproc/block/blockProcessor.go:263]: indexOfFirstTxProcessed amd indexOfLastTxProcessed

**Review by iulianpascalau — DISMISSED** [review]

File: process/elasticproc/block/blockProcessor.go [hunk 8]
```diff
@@ -227,17 +230,53 @@ func (bp *blockProcessor) getEncodedMBSHashes(body *block.Body) []string {
...
+	for idx, txHash := range txsHashes {
+		tx, found := txsMap[string(txHash)]
+		if !found {
+			log.Warn("blockProcessor.extractExecutionOrderIndicesFromPool cannot find tx in pool", "txHash", hex.EncodeToString(txHash))
+			continue
```
Comment (SebastianMarian) [inline process/elasticproc/block/blockProcessor.go]: Here is a `continue` without setting something in executionOrderTxsIndices slice. Therefore the value will remain 0 as indice, it is ok?

File: process/elasticproc/block/blockProcessor.go [hunk 9]
```diff
@@ -288,3 +327,26 @@ func createShardIdentifier(shardID uint32) uint32 {
...
+	case nodeBlock.TxBlock:
+		return pool.Txs
+	case nodeBlock.InvalidBlock:
+		return pool.Invalid
+	case nodeBlock.RewardsBlock:
+		return pool.Rewards
+	case nodeBlock.SmartContractResultBlock:
+		return pool.Scrs
+	default:
+		return map[string]coreData.TransactionHandlerWithGasUsedAndFee{}
```
Comment (SebastianMarian) [inline process/elasticproc/block/blockProcessor.go:346]: return `make(map[string]coreData.TransactionHandlerWithGasUsedAndFee{})`, otherwise it will panic in L262 ?

**Review by AdoAdoAdo — DISMISSED** [review]

File: process/elasticproc/block/blockProcessor.go (hunk 7 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor.go:263]: renamed

File: process/elasticproc/block/blockProcessor.go (hunk 8 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor.go]: added a const for `notFound` for this case and set the value with ` -2`

File: process/elasticproc/block/blockProcessor.go (hunk 9 above)
Comment (miiu96) [reply process/elasticproc/block/blockProcessor.go:346]: done.

**Review by iulianpascalau — APPROVED** [review]

File: process/elasticproc/block/blockProcessor.go (hunk 8 above)
Comment (SebastianMarian) [reply process/elasticproc/block/blockProcessor.go:268]: ok

**Review by SebastianMarian — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]
